src/mctp-bench.c

```diff
@@ -0,0 +1,413 @@
+#define _XOPEN_SOURCE 700
+#include <bits/time.h>
+#include <err.h>
+#include <errno.h>
+#include <poll.h>
+#include <signal.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/poll.h>
+#include <sys/socket.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "mctp.h"
+
+struct mctp_bench_send_args {
+        mctp_eid_t eid;
+        size_t len;
+        int net;
+};
+
+struct msg_header {
+        uint16_t magic;
+        uint32_t seq_no;
+};
+
+struct mctp_stats {
+        size_t total_received_len, curr_packet_len;
+        uint32_t prev_seq_no;
+        double elapsed_time;
+        unsigned long msgs_dropped, msg_count, invalid_payloads;
+};
+
+struct recv_ctx {
+        struct mctp_stats stats;
+        struct timespec start_time, current_time;
+        unsigned char *buf;
+        volatile bool started_recv_flag;
+        int sd;
+};
+
+static const size_t MSG_HEADER_LEN = sizeof(struct msg_header);
+static const size_t MAX_LEN = 64 * 1024 - 1;
+static const uint32_t SEQ_START = UINT32_MAX - 5;
+static const uint16_t MAGIC_VAL = 0xbeca;
+static const int DEFAULT_NET = 1;
+static const int DEFAULT_SECONDS_INTERVAL = 10;
+
+static float get_throughput(float total_len, float elapsed_time)
+{
+        return total_len / (elapsed_time * 1024);
+}
+
+static void print_stats(struct recv_ctx *recv_ctx)
+{
+        float throughput =
+            get_throughput((float)recv_ctx->stats.total_received_len,
+                           (float)recv_ctx->stats.elapsed_time);
+        printf("Throughput: %.2f kB/s | Recevied: %lu msgs | "
+               "Dropped: %lu msgs | "
+               "Invalid: %lu msgs\n",
+               throughput, recv_ctx->stats.msg_count,
+               recv_ctx->stats.msgs_dropped, recv_ctx->stats.invalid_payloads);
+}
+
+static int get_elapsed_time(struct recv_ctx *recv_ctx)
+{
+        return (recv_ctx->current_time.tv_sec - recv_ctx->start_time.tv_sec) +
+               (recv_ctx->current_time.tv_nsec - recv_ctx->start_time.tv_nsec) /
+                   1e9;
```

Comment:
Should this return a float? Currently it'll truncate to seconds.

Reply:
Changed function and calculation to use floating point number to get higher precision for elapsed time.